### PR TITLE
Curators should be able to add/remove reviewers inside their collections

### LINKED ISSUE #642: Curators should be able to add/remove reviewers inside their collections
As a curator, I want to add/remove users as reviewers in collections for which I'm a curator so that I can control which users can review for each collection.

Required by #665

### LINKED ISSUE #706: Add Button to Trigger Add Reviewer Modal for #642
Child of #642

### LINKED ISSUE #707: Implement User Search in Modal for #642
Reuse contributor search

Child of #642

### LINKED ISSUE #708: Authorization of Curator Role for #642
Child of #642

### PR DESCRIPTION
**Pull request by jbarlow96**
The purpose of this PR is to allow curators to view and manage reviewers within their collection.

Screenshots
Screen Shot 2019-03-14 at 10 50 09 PM

Screen Shot 2019-03-15 at 6 54 49 PM

Screen Shot 2019-03-15 at 12 57 45 AM

Screen Shot 2019-03-23 at 8 10 37 PM

Needs https://github.com/Cyber4All/user-service/pull/122
Needs https://github.com/Cyber4All/CLARK-Gateway/pull/73#pullrequestreview-214301520

Closes #642 
Closes https://github.com/Cyber4All/clark-client/issues/643
Closes https://github.com/Cyber4All/clark-client/issues/644
Closes #706 
Closes #707 
Closes #708

### CODE REVIEW

File: src/app/admin/pages/reviewers/reviewers.component.scss [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+.search__input {
+  width: 90%;
+  border: none;
+  outline: none;
+}
+
+.search__input::placeholder {
+  color: $placeholder;
```
Comment (nrwinner) [inline src/app/admin/pages/reviewers/reviewers.component.scss]: If I recall correctly, the `$placeholder` variable is sampled directly from the Chrome placeholder color value. Why is this here?

File: src/app/admin/pages/reviewers/reviewers.component.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      changes.forEachItem(x => {
+        this.selectedReviewers.push(x.item.username);
+      });
+    }
+  }
+
+  async fetchReviewers() {
+    console.log('HERE');
+    this.reviewers = await this.user.fetchReviewers(this.activeCollection);
+    console.log('EH', this.reviewers);
```
Comment (nrwinner) [inline src/app/admin/pages/reviewers/reviewers.component.ts]: Remove devlogs

File: src/app/admin/pages/reviewers/reviewers.component.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    if (changes) {
+      this.selectedReviewers = [];
+      changes.forEachItem(x => {
+        this.selectedReviewers.push(x.item.username);
+      });
+    }
+  }
+
+  async fetchReviewers() {
+    console.log('HERE');
```
Comment (nrwinner) [inline src/app/admin/pages/reviewers/reviewers.component.ts]: Remove devlogs

File: src/app/admin/pages/reviewers/reviewers.component.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
+import { Component, OnInit, OnDestroy, Output, EventEmitter, Input, IterableDiffers, IterableDiffer, DoCheck  } from '@angular/core';
```
Comment (nrwinner) [inline src/app/admin/pages/reviewers/reviewers.component.ts]: Is there a reason this component was copied over and slightly modified instead of made more generic and reused?

**Comment by nrwinner** [comment]
Good work on this so far. There appear to be some inconsistencies between the UI and the mocks, namely the padding in the input field and the dropdown needs to be increased. Additionally, the users icon should be the lighter regular variant (`far fa-users` instead of `fas fa-users`) and there needs to be a little more spacing between the users icon the the empty text. I've attached that section of the mocks for reference.

Screen Shot 2019-03-15 at 11 06 45 AM

File: src/app/admin/pages/reviewers/reviewers.component.scss (hunk 1 above)
Comment (DonSeannelly) [reply src/app/admin/pages/reviewers/reviewers.component.scss]: @nrwinner maybe we should add a comment explaining that. I had no idea that was the case. Also, does that mean other browsers have a different color? Do we want that to fluctuate, or be consistent?

File: src/app/admin/pages/reviewers/reviewers.component.html [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+              <div class="prompt__text">
+                  Your search results will <br> appear here
+              </div> 
+            </div>
+            <div *ngIf="loading" class="container__prompt__loading" >
+                <div class="loading-template">
+                    <i class="far fa-spinner-third fa-spin"></i>
+                    Loading Reviewers...
+                  </div>
+              </div>
```
[the remark this answers is not in the document]
Comment (DonSeannelly) [reply src/app/admin/pages/reviewers/reviewers.component.html]: We should get prettier activated in this repo to avoid this altogether.

File: src/app/admin/pages/reviewers/reviewers.component.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    this.searchResults = [];
+    this.loading = false;
+    this.showDropdown = false;
+  }
+
+  navigateToUserObjects(username: string) {
+    this.router.navigate(['admin/learning-objects'], { queryParams: { username } });
+  }
+
+  OnDestroy() {
```
Comment (DonSeannelly) [inline src/app/admin/pages/reviewers/reviewers.component.ts]: I believe this should be `ngOnDestroy()`

**Comment by nvisal1** [comment]
> To recap our conversation: We want to take the search dialog out of the ReviewersComponent, and make it generic to be just a user search that has an output event emitter with the user that was clicked.

I will create a wrapper compone

File: src/app/admin/pages/reviewers/reviewers.component.scss (hunk 1 above)
Comment (nvisal1) [reply src/app/admin/pages/reviewers/reviewers.component.scss]: What do you recommend I replace the use of `$placeholder` with?

Comment (nrwinner) [reply src/app/admin/pages/reviewers/reviewers.component.scss]: @DonSeannelly @nvisal1 Browsers do have slightly different placeholder values (they're all light gray though). That variable was intended for color-matching icons to input placeholders as closely as possible. That rule can be removed since we're taking the input for an element that's already styled by the browser and manually restyling it to essentially the same thing.

Comment (DonSeannelly) [reply src/app/admin/pages/reviewers/reviewers.component.scss]: @nrwinner we should definitely add a comment to the stylesheet on that. I had no idea that was the purpose, and I've definitely used it for things other than icons.

Comment (nvisal1) [reply src/app/admin/pages/reviewers/reviewers.component.scss]: Removing this rule makes the placeholder text much darker
Screen Shot 2019-03-25 at 7 16 15 AM
Thoughts?

File: src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  ngOnInit() {
+      // subscribe to the search input and fire search after debounce
+      this.userSearchInput$.pipe(
+        debounceTime(650),
+        takeUntil(this.destroyed$)
+      ).subscribe((val: string) => {
+        this.findUser(val.trim());
+      });
+
+      // TODO is there a better way to do this?
```
Comment (DonSeannelly) [inline src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts]: I think it makes sense to have these as two separate subscriptions, since they are in fact two different observers doing two different things.

File: src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  /**
+   * Queries the UserService with query text and sets the searchResults variable
+   *
+   * @memberof UserSearchWrapperComponent
+   */
+  findUser(query: string) {
+    if (query && query !== '') {
+      this.user.searchUsers(query).then((results: User[]) => {
+        // remove current user from results
+        for (let i = 0; i < results.length; i++) {
```
Comment (DonSeannelly) [inline src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts]: This for loop could be replaced by using `Array.filter` from the Array prototype.

File: src/app/admin/pages/learning-objects/learning-objects.component.ts [hunk 9]
```diff
@@ -31,24 +31,26 @@ export class LearningObjectsComponent implements OnInit, OnDestroy {
...
 
   ngOnInit(): void {
-    this.route.queryParams
-      .pipe(
-        takeUntil(this.componentDestroyed$),
-      )
-      .subscribe(u => this.getUserLearningObjects(u.username));
+    this.route.queryParams.subscribe(params => {
+      const username = params['username'];
+      if (typeof(username) !== 'undefined' && username !== null) {
```
Comment (DonSeannelly) [inline src/app/admin/pages/learning-objects/learning-objects.component.ts]: Can we be sure to scrub the typeof undefined statements from this PR? I think we had a conversation or two about this before. 
```
if (username == undefined) { ... }
```
does the same thing and is much less verbose.

File: src/app/admin/pages/learning-objects/learning-objects.component.ts [hunk 10]
```diff
@@ -57,6 +59,11 @@ export class LearningObjectsComponent implements OnInit, OnDestroy {
       });
   }
 
+  reload(event) {
```
Comment (DonSeannelly) [inline src/app/admin/pages/learning-objects/learning-objects.component.ts]: Looks like this was here just for logging in dev

File: src/app/admin/pages/learning-objects/learning-objects.component.ts (hunk 9 above)
Comment (gshaw1997) [reply src/app/admin/pages/learning-objects/learning-objects.component.ts]: @DonSeannelly Is `if (username == undefined) { ... }` or `if (username == null) { ... }` the better option for this check?

Comment (DonSeannelly) [reply src/app/admin/pages/learning-objects/learning-objects.component.ts]: @gshaw1997 I think it is entirely personal opinion. Both are bottom values so the same operation is done under the hood.

Personally, I favor the if-null approach since null is fewer characters to type. I think that we should just pick one as a team and stick with it.

**Comment by jbarlow96** [comment]
Screen Shot 2019-03-27 at 11 25 27 AM

**Comment by nvisal1** [comment]
The wording on the new modal should be changed. Saying "you can't undo this" might not make sense because the curator can go back and add that user again.

**Comment by jbarlow96** [comment]
Screen Shot 2019-03-29 at 9 24 26 AM

File: src/app/core/user.service.ts [hunk 11]
```diff
@@ -54,6 +54,91 @@ export class UserService {
...
+  async removeMember(
+    collection: string,
+    memberId: string,
+    role: any,
+  ): Promise<void> {
+    await this.http.request(
+        'delete',
+        USER_ROUTES.REMOVE_COLLECTION_MEMBER(collection, memberId),
+        {
+          body: role,
```
Comment (nvisal1) [inline src/app/core/user.service.ts]: This route no longer requires a role.

File: src/app/admin/pages/users/users.component.ts [hunk 12]
```diff
@@ -27,6 +44,42 @@ export class UsersComponent implements OnInit {
...
+    this.removeReviewerId = reviewerId;
+    this.displayRemoveReviewerModal = true;
+  }
+
+  /**
+   * Removes the reviewer access from a user in a collection
+   */
+  async removeReviewer() {
+    this.displayRemoveReviewerModal = false;
+    await this.user.removeMember(this.activeCollection, this.removeReviewerId, {role: 'reviewer'} );
```
Comment (nvisal1) [inline src/app/admin/pages/users/users.component.ts]: You removed the role parameter from this function. This will throw an error.

File: package.json [hunk 13]
```diff
@@ -9,8 +9,7 @@
     "build": "ng build",
     "test": "jest --verbose  --coverage",
     "lint": "ng lint",
-    "app": "cd .. cd ",
-    "e2e": "cypress open"
+    "app": "cd .. cd "
```
Comment (gshaw1997) [inline package.json]: Whut?
